zkml/docs/src/softmax.md

```diff
@@ -0,0 +1,80 @@
+# Softmax
+
+## Description of Layer
+
+Formally Softmax is a linear operator that maps $`\mathbb{R}^{n}`$ to $`\mathcal{U}_{n}`$, where $`\mathcal{U}_{n}`$ is the space of probability distributions for discrete random variables with $`n`$ possible outcomes. Importantly for a vector $`v \in \mathbb{R}^{n}`$ it maps each component of $`v`$ to the interval $`(0,1)`$ and satisfies the sum of the components is $`1`$. The definition of the map is
+
+$$ \begin{align*} \mathrm{Softmax}(v):& \mathbb{R}^{n}\rightarrow \mathcal{U}_{n} \\
+&v \mapsto \left(\frac{\exp(v_{i})}{\sum_{j=1}^{n}\exp(v_{j})}\right)_{i=1}^{n}\end{align*} $$
+
+The Softmax layer is used to inside a transformer to map a tensor of weighted values to a probability distribution. The QKV layer produces three tensors:
+ 
+ - The query $` Q \in \mathbb{R}^{h}`$
+ - The keys $` K \in \mathbb{R}^{s\times h}`$
+ - The values $` V \in \mathbb{R}^{s \times h}`$
+
+ where $` h`$ is the head dimension and $`s`$ is the sequence length. Softmax is employed on $`QK^{\intercal}\in \mathbb{R}^{s}`$, namely we calculate 
+
+ $$ \begin{align*} \mathrm{Softmax}\left(\frac{QK^{\intercal}}{\sqrt{h}}\right) = \left(\frac{\exp(\frac{QK^{\intercal}_{i}}{\sqrt{h}})}{\sum_{j=1}^{s}\exp(\frac{QK^{\intercal}_{j}}{\sqrt{h}})}\right)_{i=1}^{s}.\end{align*} $$
+
+ ### Shift Invariance
+
+ One of the key properties of Softmax which we make use of to both stabilise inference and to assist with proving is that is invariant under shifts. That is given a vector $`v\in\mathbb{R}^{d}`$ and a constant vector $`\emph{c}\in\mathbb{R}^{d}`$ then 
+
+ $$ \begin{align*}  \mathrm{Softmax}(v - c)_{i} &= \frac{\exp(v_{i}-c)}{\sum_{j=1}^{d}\exp(v_{j}-c)} \\ &=\frac{\exp(-c)\cdot\exp(v_{i})}{\exp(-c)\cdot\sum_{j=1}^{d}\exp(v_{j})}\\ &=\frac{\exp(v_{i})}{\sum_{j=1}^{d}\exp(v_{j})} \\ &= \mathrm{Softmax}(v)_{i}.\end{align*} $$
+
+ ## Proving the Layer
+
+ Due to the highly non-linear nature, and the fact that the $`\exp`$ function has no natural analogue over a finite field $`\mathbb{F}`$, we employ lookups to prove correct execution of softmax. The technique is largely based on that proposed in [zkLLM][1], the notable differences being due to Deep Proves use of the Goldilocks field we perform quantisation differently. 
+
+  ### Exponential Tables
+
+ In order to reuse lookup tables between softmax layers we assume that the scaling factor of the quantised inputs is the same everytime. For this we pick a suitable power of $`2`$ (in our case $`2^{24}`$) and consider our quantised values $`q`$ to be such that if $`x\in\mathbb{R}`$ then $`2^{24} \cdot x = q`$. 
+
```

Comment:
As mentioned also by @nikkolasg, I would maybe expand here with the details we discussed on the call, which were super-helpful to understand what is done. Namely, the fact that to get a quantized value `q` s.t. `q = x * 2^24`, we need to find the multiplier `y` such that `q*y = x * 2^24`, given the actual scaling factor of input `q`.